### ISSUE TITLE
Jakarta EE 11 must support certification under Java SE 17 in addition to Java SE 21

### ISSUE THREAD

**Issue by starksm64** [jakartaee/platform#820]
**Is your feature request related to a problem? Please describe.**
The current Java SE 21 mandate is overreaching in terms of a requirement if EE 11 implementations cannot even boot using the non-normative API component specification artifacts under Java SE 17.

**Describe the solution you'd like**
While Java SE 21 can be a mandated runtime for certification, there still must be support for Java SE 17 as a valid runtime for certification.

**Describe alternatives you've considered**
We will have to fork Jakarta EE 11 specification projects that compile with a target language level of Java SE 21 to either produce a lower target version or a MR version of the artifact. If a TCK requires a Java SE 21 runtime it is unclear what approach can be taken. This will impact our ability to provide CIs for Jakarta Validation and Jakarta CDI.

**Additional context**

The following New Relic survey of Java version adoption is consistent with our user base adoption profile:
https://newrelic.com/resources/report/2023-state-of-the-java-ecosystem

![image](https://github.com/jakartaee/platform/assets/332210/097ddfb7-fc0c-424d-b272-1d89bf811844)

Java SE 21 is not even in the picture because of its Q4 2023 release date. The survey shows that most adoption is skewed toward the current LTS(-1) version. A best case scenario is that Java SE 21 sees a large first year adoption rate, but our expectation is that similar to this survey, this still leaves more than 90% of our userbase still be on an LTS release prior to Java SE 21.

**Comment by starksm64** [comment]
Another data point from our Quarkus users, who do look further ahead than our Jakarta EE customers is the following distribution of what Java SE a build targets:
![image](https://github.com/jakartaee/platform/assets/332210/9e69d788-ce16-4452-a9c2-095a3975ee55)

while there is clear adoption of Java SE 21, the majority is on 17 or lower and there is no way we would restrict to Java SE 21 or higher for any release in 2024.

**Comment by starksm64** [comment]
Another data point from any developer using VSCode and Java that opts into telemetry tracking shows the following distribution of build targets:

![vscode_java_source_min](https://github.com/jakartaee/platform/assets/332210/1b80950a-0ea8-47d2-a526-2ff48c54a051)


Again, there is adoption of Java SE 21, but it is far from the preferred target with all three previous Java SE LTS releases having larger usage.

**Comment by hantsy** [comment]
That means the virtual thread will be optional(or not available) in Jakarta EE 11?

**Comment by jamezp** [comment]
> That means the virtual thread will be optional(or not available) in Jakarta EE 11?

The only specification wanting use of virtual threads is Jakarta Concurrency and it's already optional. See https://github.com/jakartaee/concurrency/issues/268

**Comment by edburns** [comment]
Tracking at https://dev.azure.com/jakarta-ee-azdo/jakarta-ee-azdo/_workitems/edit/83

**Comment by arjantijms** [comment]
> this still leaves more than 90% of our userbase still be on an LTS release prior to Java SE 21.

The one thing I don't understand is the assumption that users who are on a LTS release prior to Java SE 21 will not also be on an LTS release prior to Jboss EAP 9.0.

It's clear that a given amount of companies are not on JDK 21, but where does the assumption come from that updating the JDK to 21 is not possible for those companies, but updating Jboss EAP to 9.0 is doable?

(Ignoring for the moment that Jboss EAP 8.0 is not even released, so upgrading to Jboss EAP 9.0 may not be an issue until 2025/2026?)

**Comment by arjantijms** [comment]
> Again, there is adoption of Java SE 21, but it is far from the preferred target with all three previous Java SE LTS releases having larger usage.

But take a look at the adoption of Jakarta EE 11. That one is at the moment zero (since, of course, Jakarta EE 11 has not been released yet). So JDK 21 is already ahead of Jakarta EE 11.

Likely, by the time Jakarta EE 11 will start to see significant adoption rates, JDK 21 will be old news and JDK 25 will have been released.

Especially with docker deployments, the JDK and the AS are typically upgraded together (from my personal experience at least).

**Comment by jamezp** [comment]
> > this still leaves more than 90% of our userbase still be on an LTS release prior to Java SE 21.
> 
> The one thing I don't understand is the assumption that users who are on a LTS release prior to Java SE 21 will not also be on an LTS release prior to Jboss EAP 9.0.
> 
> It's clear that a given amount of companies are not on JDK 21, but where does the assumption come from that updating the JDK to 21 is not possible for those companies, but updating Jboss EAP to 9.0 is doable?
> 
> (Ignoring for the moment that Jboss EAP 8.0 is not even released, so upgrading to Jboss EAP 9.0 may not be an issue until 2025/2026?)

Red Hat has more than just JBoss EAP FWIW. There is also Quarkus which has plans to support Java SE 17 as well. While they won't be a full Jakarta EE 11 container, they will want to implement/support some of the individual specifications. Having specifications, specifically Jakarta REST 4.0, have a minimum of Java 21 means either we need to fork the specification API and compile it down to Java SE 17 or the specification can't be used. IMO this is bad in multiple ways.

Note the only two specs using Java 21 bytecode are Jakarta REST and Jakarta Concurrency. The latter is the only one that has a real reason for requiring that. There is no technical reason why any of the specification API's need to compile with Java 21 bytecode.

This doesn't mean users won't use Java 21 or can't use Java 21. It just means they could choose Java 17 or Java 21. I feel it's better, especially given no technical reason, for the user to decide what works best for them. Not have the platform tell them what's best for them.

**Comment by jamezp** [comment]
> We are talking about minimum requirements for specs and platform.
> 
> Specs can decide, that they will support older JDKs like 17 or even 11, they are not limited. They just need to support also the platform's one, e.g. if a spec supports 11 and works well in 21, everything is fine.

Correct. The issue specifically for us is the Jakarta REST spec has chosen to require 21 and a request to downgrade has been vetoed.

> 
> On the other hand, if the decision will be CHANGED to base Jakarta EE 11 platform on JDK 17, it will require limit specs <=17. Implementations using JDK 21 features won't pass TCK on JDK 17, e.g. cannot be certified.
> 
> An option, that the certification can be done on 17 OR 21 seems to solve this situation as well.

+1, we've got no issue with a requirement to pass both a 17 and 21 TCK. Our issue is having API's require Java 21 for no technical reason.

**Comment by arjantijms** [comment]
> Red Hat has more than just JBoss EAP FWIW.

I know, but JBoss EAP was the one that was repeatedly mentioned. Specifically since this product appeals to customers who are generally slower to adopt the latest technology. WildFly users, if I'm not mistaken, would have far less issue with using new technology. 

> they will want to implement/support some of the individual specifications. Having specifications, specifically Jakarta REST 4.0

By why would the Quarkus users who are slow to adopt new technology, want to use Jakarta REST 4.0? That's the part you still haven't explained really. Jakarta REST 4.0 is new, and contains breaking changes. So the type of user who avoids JDK 21, because it's so new and may break things, would also avoid Jakarta REST 4.0 right?

**Comment by jamezp** [comment]
> > Red Hat has more than just JBoss EAP FWIW.
> 
> I know, but JBoss EAP was the one that was repeatedly mentioned. Specifically since this product appeals to customers who are generally slower to adopt the latest technology. WildFly users, if I'm not mistaken, would have far less issue with using new technology.

I've seen more WildFly users using Java 17 than Java 21 for sure. I don't have data on it by any means, but from my experience working on the project, that's what I've seen.

> 
> > they will want to implement/support some of the individual specifications. Having specifications, specifically Jakarta REST 4.0
> 
> By why would the Quarkus users who are slow to adopt new technology, want to use Jakarta REST 4.0? That's the part you still haven't explained really. Jakarta REST 4.0 is new, and contains breaking changes. So the type of user who avoids JDK 21, because it's so new and may break things, would also avoid Jakarta REST 4.0 right?

I don't think Quarkus users are slow to adopt new technology. However, I cannot speak directly to this as I'm not as involved with that project. Quarkus itself is likely considered fairly new really.

**Comment by arjantijms** [comment]
> Our issue is having API's require Java 21 for no technical reason.

There's a very good technical reason, and that's to force all implementations to be at least Java 21 compatible. As a library vendor you know that if you want to target ALL Jakarta REST implementations, you can use JDK 21 for your product. If the compatibility requirements are weakened to also allow JDK 17, then a vendor can no longer use JDK 21 features in their product.

The API would probably still compile with JDK 8, and likely even JDK 7 or JDK 6. But that would allow for JDK 6 implementations, which throws away guarantees about which JDK vendors can use, and additionally drags the entire ecosystem down.

**Comment by jamezp** [comment]
> > The issue specifically for us is the Jakarta REST spec has chosen to require 21 and a request to downgrade has been vetoed.
> 
> To be clear on this, I personally would not veto the usage to downgrade. I mentioned this in the platform call; if the entire group choses to prefer JDK 17 as a minimum, with optional support for JDK 21 via certification, I would accept that. OmniFish would accept it too.
> 
> It's just that I personally don't agree with all the arguments I've seen against a JDK 21 only release. But people sometimes disagree, and we can agree to disagree then.

Agreed :) And I recall you even said that on the rest-dev mailing list too. You were a -0 vote, but not a veto. Seemed reasonable to me :)

**Comment by arjantijms** [comment]
> I don't think Quarkus users are slow to adopt new technology.

If they are not, what is the issue with JDK 21 then?

It will take several months still to release Jakarta EE 11, and after that maybe several months or longer until Quarkus supports EE 11. I remember Quarkus wasn't super fast with adopting Jakarta EE 10 either. See e.g. https://quarkus.io/blog/our-bumpy-road-to-jakarta-ee-10/

So for a potential Quarkus release including Jakarta REST 4.0, we may be looking at what, mid 2025 realistically?

What type of Quarkus user, who as you say is not afraid of using new technology, would be afraid to use JDK 21 mid 2025, but would not be afraid to use the much more "dangerous" Jakarta REST 4.0?

**Comment by jamezp** [comment]
> > Our issue is having API's require Java 21 for no technical reason.
> 
> There's a very good technical reason, and that's to force all implementations to be at least Java 21 compatible. As a library vendor you know that if you want to target ALL Jakarta REST implementations, you can use JDK 21 for your product. If the compatibility requirements are weakened to also allow JDK 17, then a vendor can no longer use JDK 21 features in their product.
> 

I guess I don't understand this. Application servers consume libraries, and even expose them to end users, all the time that are compiled to lower byte code. Think pretty much any logging facade. That doesn't mean they don't work with new JVM's at all. It just means they work on both older and newer JVM's.

The one case I can see is if an implementation decides they will be compiled down to Java SE 21 byte code, then the community might complain. However, it seems that should be left for the implementation to decide.

> The API would probably still compile with JDK 8, and likely even JDK 7 or JDK 6. But that would allow for JDK 6 implementations, which throws away guarantees about which JDK vendors can use, and additionally drags the entire ecosystem down.

If it's tested on Java 21, that seems fine. You can easily guarantee this by testing on new JVM's. We do it for WildFly and various other projects for Jakarta EE 10 now.

**Comment by jamezp** [comment]
> If they are not, what is the issue with JDK 21 then?

The pie chart added to this issue is from proprietary Quarkus data. It clearly shows that Java 17 is by far the most common used, e.g. latest LTS -1. If they want to retain users, they need to continue to support Java 17.

> 
> It will take several months still to release Jakarta EE 11, and after that maybe several months or longer until Quarkus supports EE 11. I remember Quarkus wasn't super fast with adopting Jakarta EE 10 either. See e.g. https://quarkus.io/blog/our-bumpy-road-to-jakarta-ee-10/
> 
> So for a potential Quarkus release including Jakarta REST 4.0, we may be looking at what, mid 2025 realistically?
> 
> What type of Quarkus user, who as you say is not afraid of using new technology, would be afraid to use JDK 21 mid 2025, but would not be afraid to use the much more "dangerous" Jakarta REST 4.0?

I cannot speak for Quarkus. I'm not on that team, so I don't fully know their plans or road map with adopting the new Jakarta EE 11 specifications.

**Comment by arjantijms** [comment]
> The pie chart added to this issue is from proprietary Quarkus data. It clearly shows that Java 17 is by far the most common used, e.g. latest LTS -1. If they want to retain users, they need to continue to support Java 17.

I think, with all respect, that the conclusion is not correct.

The data shows that for the Jakarta EE 10/REST 3.1 compatible Quarkus, Java 17 is by far the most commonly used.

It says nothing about how eager any of those people are to move to Jakarta EE 11/REST 4.0, and when they are eager / daring (bold) enough to move to REST 4.0, whether they would also not move to JDK 21 as well.

It comes down again and again on the assumption that people would not care that much about changing the 3.1 in the POM to 4.0 where it concerns REST, but that they would have big issues with changing the 17 to 21 in the <release> tag of that same POM. Logically that doesn't make sense. Changing the 3.1 to 4.0 entails a far bigger risk than changing 17 to 21.

Until so far I haven't seen anyone come up with any prove that users prefer such unbalanced updates. In fact, if anything the exact contrary is shown. If users preferred unbalanced updates, we would see a lot of users for Jakarta EE 10/REST 3.1 compatible Quarkus on JDK 11 now, but they are not all massively on 11. They are on 17.

**Comment by arjantijms** [comment]
p.s.

> It clearly shows that Java 17 is by far the most common used, e.g. latest LTS -1.

Note that it's the latest LTS-1 *now*. But it was LTS 0 when Jakarta EE 10 was released.

**Comment by NottyCode** [comment]
> > Our issue is having API's require Java 21 for no technical reason.
> 
> There's a very good technical reason, and that's to force all implementations to be at least Java 21 compatible. As a library vendor you know that if you want to target ALL Jakarta REST implementations, you can use JDK 21 for your product. If the compatibility requirements are weakened to also allow JDK 17, then a vendor can no longer use JDK 21 features in their product.
> 

I do not agree that this is a very good technical reason. I think a very good technical reason for compiling to Java SE 21 is that you make use of a Java API or language feature that is only available in Java SE 21. Forcing implementations to use Java SE 21 is more of a political choice than a technical one. It isn't even a very good technical solution to the stated problem here because as Scott points out an implementation can just take the source and recompile it for Java 17, I wouldn't even claim that was a fork.

Even if we support the API compiling back to Java SE 17 as Scott is requesting it doesn't by itself preclude a different implementation deciding to use and require Java SE 21. WebSphere Liberty supports Java EE 7, but it doesn't allow it to be run on Java SE 7 anymore.

> Until so far I haven't seen anyone come up with any prove that users prefer such unbalanced updates.

I suspect it is unlikely anyone will be able to prove to you that someone would want this. I speak to a lot of people who use Jakarta EE, but because those conversations are confidential in nature it is hard to produce evidence from other than generalized anecdotes. 

One generalization is that there is a desire from some users to run Java EE 7 applications on Java SE 17. The reason for this is because they don't want to have to migrate to Jakarta EE 8 or 10, but they are concerned that Java SE 8 end of support is looming, or they think upgrading to Java SE 17 will give them an automatic performance boost.

Another generalization is that there are a lot of enterprises where the EE version choice is down to the application team, but the Java SE version decision is made by an operations or platform team. If the operations team have "standardized" on Java SE 17, but there is a really cool feature in Jakarta EE 11 then an application team may wish to use it, but getting an upgrade to Java SE 17 to 21 might be a significant challenge within their enterprises processes.

My point here isn't to say I have better data, but just to observe that I too have experience talking to users of EE where having a wider support window could be advantageous.

The counter point you have made is also fair which is that by the time we release and people adopt Jakarta EE 11 it is possible that they will all be on Java SE 21, but the status quo here will make that a certainty. Oracle's end of support for Java SE 17 is September 2029, Eclipse Temurin's [release roadmap](https://adoptium.net/support/#_release_roadmap) has releases of Java SE 17 until October 2027 so both will be around for quite a while after EE 11 is due to be released.

**Comment by jamezp** [comment]
> 
> I think, with all respect, that the conclusion is not correct.
> 
> The data shows that for the Jakarta EE 10/REST 3.1 compatible Quarkus, Java 17 is by far the most commonly used.
> 
> It says nothing about how eager any of those people are to move to Jakarta EE 11/REST 4.0, and when they are eager / daring (bold) enough to move to REST 4.0, whether they would also not move to JDK 21 as well.
> 
> It comes down again and again on the assumption that people would not care that much about changing the 3.1 in the POM to 4.0 where it concerns REST, but that they would have big issues with changing the 17 to 21 in the `<release>` tag of that same POM. Logically that doesn't make sense. Changing the 3.1 to 4.0 entails a far bigger risk than changing 17 to 21.
> 
> Until so far I haven't seen anyone come up with any prove that users prefer such unbalanced updates. In fact, if anything the exact contrary is shown. If users preferred unbalanced updates, we would see a lot of users for Jakarta EE 10/REST 3.1 compatible Quarkus on JDK 11 now, but they are not all massively on 11. They are on 17.

Maybe I'm not explaining my position correctly. All I'm saying is there is no technical reason to compile to Java 21 byte code. No API's are used that are only in Java 21. The only reason it's being done is because "we want to", which is not a technical reason at all. It's not that I don't understand the desire to want it to be used on the latest LTS and the latest features. My default JDK is Java 21 now and I run pretty much everything on it :)

Maybe users would be fine with it requiring Java 21. Maybe they won't. The data of what is currently being used is all we can go on. We can only base our decisions on feedback from what customers and users are doing. We had, I think two, responses on the rest-dev from users who would prefer the API to work on Java 17.

Not to bring it up here too, but the Jakarta REST spec was originally passed with a minimum of Java SE 17. The way the bump to Java 21 happened was under the "it's not an API change" rule. However, I completely disagree with that. If I have to make a change to run the API, that's a break and the API is not compatible.

**Comment by arjantijms** [comment]
> Forcing implementations to use Java SE 21 is more of a political choice than a technical one.

It gives you a guarantee, which after all is the purpose of a specification.

For instance, when JSF 2.3 made the decision to use Java 8 (scary at the time we made that decision, already a little less scary by the time JSF 2.3 was released, and not scary at all when people actually started using it), it gave vendors targeting JSF 2.3 the guarantee that their library/extension could use Java 8. E.g. PrimeFaces, OmniFaces, etc could all use JDK 8 without worrying that someone complained it would not run on their JSF 2.3 implementation.

Would JSF 2.3 not have made this decision, but would have said Java 8 or Java 7, then effectively PrimeFaces, OmniFaces etc would only be able to use Java 7, as there would be no guarantee all JSF 2.3 implementations were running on Java 8.

**Comment by arjantijms** [comment]
> Maybe I'm not explaining my position correctly. All I'm saying is there is no technical reason to compile to Java 21 byte code. No API's are used that are only in Java 21.

But then again, there's also no technical reason to compile to Java 17 or even Java 11 is there? The lowest version the REST 4.0.0-SNAPSHOT API compiles to is currently Java 9. Yet we [updated it before to JDK 11](https://github.com/jakartaee/rest/commit/c99d1286d14524ed726f0287d5c366abe7bbf3c7).

**Comment by jamezp** [comment]
> > Maybe I'm not explaining my position correctly. All I'm saying is there is no technical reason to compile to Java 21 byte code. No API's are used that are only in Java 21.
> 
> But then again, there's also no technical reason to compile to Java 17 or even Java 11 is there? The lowest version the REST 4.0.0-SNAPSHOT API compiles to is currently Java 9. Yet we [updated it before to JDK 11](https://github.com/jakartaee/rest/commit/c99d1286d14524ed726f0287d5c366abe7bbf3c7).

While the code itself doesn't require Java 17, I would argue that support for records is important. However, that could be done in other ways too with MR JAR's. It already is in Yasson for Jakarta EE 10.

**Comment by NottyCode** [comment]
> It gives you a guarantee, which after all is the purpose of a specification.

Not really because Red Hat will just take the API and recompile it for Java 17 removing any such guarantee.

> Would JSF 2.3 not have made this decision, but would have said Java 8 or Java 7, then effectively PrimeFaces, OmniFaces etc would only be able to use Java 7, as there would be no guarantee all JSF 2.3 implementations were running on Java 8.

I'm sorry but I have no idea what you think you are arguing here. Ultimately allowing the API to compile to Java 17 provides more choice. It provides implementations with more choice in that they can opt to support Java 17+ or Java 21+. It also allows a user wanting to choose Java 17 to do so by choosing an implementation that supports Java 17+. Your argument seems to be that the choice shouldn't be allowed, but it isn't clear to me why that choice should be banned.

As I said earlier a compelling technical argument would be that the API is using a Java 21 only feature or API, but from what I understand there is no such usage in any of the APIs in Jakarta EE 11.

**Comment by hantsy** [comment]
if possible to keep base API compile level at a lower version, such as Java 17. But Java compiler has support multiple versioned package at compile/runtime, for those feature requires Java 21/22 etc, use a higher version Java to compile/run it. 

For example, making the Virtual Thread feature are only available in Java 21 or above, when using it in a Java 17 runtime, throws exceptions or prints warning for developers, and disables it at all.

**Comment by arjantijms** [comment]
> I'm sorry but I have no idea what you think you are arguing here.

I'm sorry that I failed in making the point more clear. Let me try again in a more abstract way. 

If JDK 21 is the minimum, I know my code can use at least JDK 21. If JDK 17 is the minimum, I know my code can use at least JDK 17.

If I want to use JDK 21, and be compatible with all implementations, I'd like the minimum to be JDK 21.

**Comment by arjantijms** [comment]
>  It also allows a user wanting to choose Java 17 to do so by choosing an implementation that supports Java 17+. 

But this boils down again to the question that was asked several times, and still hasn't been fully answered; why would a user want that?

We could state, and I know I'm exaggerating here, but just for the sake of argument, that a user may want to choose a Servlet 3 implementation to run on, so nothing in Jakarta EE 11 can take advantage of anything from Servlet 3+ yet.

The example of an unbalanced upgrade that was provided is that the JDK was updated since support was running out, so they'd like older versions of EE to run on newer JDKs. The other way around.

But again, why would someone be afraid to update the JDK to 21, but would just update to the far more risky REST 4.0 without giving it a second thought?

And also again, the early adaptors of Jakarta EE 11 will be those people adapting it, well, early. Early adopters are not known for avoiding JDK updates. When the mass will take a look at adopting Jakarta EE 11, we'll be at JDK 24, almost 25. When the slow moving customers will even take Jakarta EE 11 into consideration wouldn't we be at JDK 29 already?

For historical evidence, look at Jakarta EE 10. The early adopters had no issue with JDK 17. Only around now the mass starts to slowly migrate to Jakarta EE 10, and today JDK 17 is absolutely not a problem. In a year or two from now the slower customers might be looking at Jakarta EE 10, and then surely JDK 17 is not a problem anymore.

**Comment by starksm64** [comment]
Requiring certification under Java SE 21 is the only thing that provides a guarantee of its usability. The non-normative specification APIs do not need to be used. Simply requiring these APIs to be compiled to a SE 21 target only restricts their usage to Java SE 21 by normal developers who are not building EE runtimes. Legacy end of life SE runtimes are removed from support. Once that happens that target can be removed, but it also does not require an update that is just busy work to increase the target level. Unless there is a binary incompatibility, there is no reason to advance binary's target SE version.

**Comment by arjantijms** [comment]
> Unless there is a binary incompatibility, there is no reason to advance binary's target SE version.

If we're really strict about that, should we not go back to Java 9 then for REST 4.0? As mentioned, that is the actual minimum level required by Jakarta REST 4.0. Or do you just mean to say the level is irrelevant at the moment, and we should leave it at whatever it is, even when the current level is not really needed either?

**Comment by hs536** [comment]
I think the discussion is becoming complicated because the discussions about the source level and the target level are mixed.
Here is a table about possible options:

|option  |Jakarta EE support|API impl src level         |user app src level    |target runtime level|
|-----   |-----             |-----                      |-----                 |-----               |
|A       |JDK17             | <= JDK17                  | <= JDK17             | JDK17              |
|B       |JDK17 and JDK21   | <= JDK17 or Multi-Release | <= JDK17 or <= JDK21 | JDK17 or JDK21     |
|C       |JDK21             | <= JDK21                  | <= JDK21             | JDK21              |

The current conflict seems to be occurring between optB and optC.
Those who say optB is desirable seem to place the most importance on the possibility of choosing the "target runtime level" from the current usage data.

Personally, I am skeptical whether "being able to choose JDK17" is a good for users.
This dual support probably results in "users who continue to use JDK17 without knowing if it will affect their system" and thus procrastinate on modernizing their system.

I believe it's better to "update to the latest unless there's a reason not to" rather than "maintain the old unless there's a reason not to". This is because old code will inevitably stop working at some point (due to reasons like API de-support).

btw my customers tend to strongly demand support for the latest JDK from the perspectives of security, performance, and stability, as long as there are no destructive incompatibilities between versions. Therefore, they are not very interested in whether EE11 supports JDK17 or not. just to add; this "latest JDK" refers to the LTS version, not the non-LTS version.

This is just my personal thought. I hope it can contribute to the discussion.

**Comment by anbusampath** [comment]
I am not able to find the source mailing-list for below messages,

> The Jakarta EE Working Group has established a policy of aligning Jakarta EE releases with Java SE LTS releases. This means that each new major version of Jakarta EE will be based on the latest LTS version of Java SE.
> These occur approximately six months after a new Java LTS version is released. For example, Jakarta EE 11 is targeted for Q1 2024, following the release of Java 21 in September 2023.

but this has been communicated in Ivar talks and blog as well - https://newsroom.eclipse.org/eclipse-newsletter/2023/august/jakarta-ee-11-next-major-jakarta-ee-update-shaping

Jakarta EE | Java SE
-- | --
11 | 21
12 | 25
13 | 29

I believe aligning with latest Java SE LTS helps more adoption as well.

**Comment by Emily-Jiang** [comment]
> I think the discussion is becoming complicated because the discussions about the source level and the target level are mixed. Here is a table about possible options:
> option 	Jakarta EE support 	API impl src level 	user app src level 	target runtime level
> A 	JDK17 	<= JDK17 	<= JDK17 	JDK17
> B 	JDK17 and JDK21 	<= JDK17 or Multi-Release 	<= JDK17 or <= JDK21 	JDK17 or JDK21
> C 	JDK21 	<= JDK21 	<= JDK21 	JDK21
> 
> The current conflict seems to be occurring between optB and optC. Those who say optB is desirable seem to place the most importance on the possibility of choosing the "target runtime level" from the current usage data.
> 
> Personally, I am skeptical whether "being able to choose JDK17" is a good for users. This dual support probably results in "users who continue to use JDK17 without knowing if it will affect their system" and thus procrastinate on modernizing their system.
> 
> I believe it's better to "update to the latest unless there's a reason not to" rather than "maintain the old unless there's a reason not to". This is because old code will inevitably stop working at some point (due to reasons like API de-support).
> 
> btw my customers tend to strongly demand support for the latest JDK from the perspectives of security, performance, and stability, as long as there are no destructive incompatibilities between versions. Therefore, they are not very interested in whether EE11 supports JDK17 or not. just to add; this "latest JDK" refers to the LTS version, not the non-LTS version.
> 
> This is just my personal thought. I hope it can contribute to the discussion.

We need to differentiate the specification of EE 11 and the implementation of EE 11. If EE 11 goes out with Java 17 & Java 21 support, an implementation can choose which Java SE they support. Without an EE 11 impl, the EE 11 APIs won't do anything. If your customer base is on the latest Java SE, the corresponding impl should just support Java 21. 

Jakarta EE is a specification and it is best to cater for different impl requirements and maximum the adoption. Then different impl can choose what works for them and make decision accordingly.

**Comment by starksm64** [comment]
> I think the discussion is becoming complicated because the discussions about the source level and the target level are mixed. Here is a table about possible options:
> 
> option	Jakarta EE support	API impl src level	user app src level	target runtime level
> A	JDK17	<= JDK17	<= JDK17	JDK17
> B	JDK17 and JDK21	<= JDK17 or Multi-Release	<= JDK17 or <= JDK21	JDK17 or JDK21
> C	JDK21	<= JDK21	<= JDK21	JDK21
> The current conflict seems to be occurring between optB and optC. Those who say optB is desirable seem to place the most importance on the possibility of choosing the "target runtime level" from the current usage data.
> 
> Personally, I am skeptical whether "being able to choose JDK17" is a good for users. This dual support probably results in "users who continue to use JDK17 without knowing if it will affect their system" and thus procrastinate on modernizing their system.
> 
> I believe it's better to "update to the latest unless there's a reason not to" rather than "maintain the old unless there's a reason not to". This is because old code will inevitably stop working at some point (due to reasons like API de-support).
> 
> btw my customers tend to strongly demand support for the latest JDK from the perspectives of security, performance, and stability, as long as there are no destructive incompatibilities between versions. Therefore, they are not very interested in whether EE11 supports JDK17 or not. just to add; this "latest JDK" refers to the LTS version, not the non-LTS version.
> 
> This is just my personal thought. I hope it can contribute to the discussion.

How are your users affected by being able to use both SE 17 and 21 as opposed to just 21?

**Comment by hs536** [comment]
> How are your users affected by being able to use both SE 17 and 21 as opposed to just 21?

As I mentioned earlier, this could lead to some users continuing to use JDK17 without checking its impact on their own systems, just to save a little cost. This will delay the modernization of the user's system.

**Comment by arjantijms** [comment]
> How are your users affected by being able to use both SE 17 and 21 as opposed to just 21?

Even though we said we shall not stand in the way if the group decides that we should go for SE 17, our users to are affected by taking away a guarantee.

Spring Boot moved to JDK 17 as a baseline in November 2022. We will now move to JDK 17 as a baseline in September 2024. Almost two years behind Spring. We didn't dare to move to JDK 17 last time around with EE 10, while Spring just did it.

We're repeating that with JDK 21 now.

**Comment by smillidge** [comment]
When drafting the steering committee goals for Jakarta EE 11 the target was set to 21 to enable specifications to use Java 21 language features in their APIs. Does this mean none have done so or plan to do so for EE 11?

**Comment by NottyCode** [comment]
@smillidge that is my understanding. Even Jakarta Concurrency which has support for creating work on virtual threads doesn't require Java SE 21 in the API and support for virtual threads is essentially optional.

**Comment by hs536** [comment]
> Even Jakarta Concurrency which has support for creating work on virtual threads doesn't require Java SE 21 in the API and support for virtual threads is essentially optional.

Even if it is optional, as long as it exists in the spec, we will need an implementation and TCK to ensure that the API works correctly.

**Comment by hs536** [comment]
is it possible to implement and test Jakarta Concurrency API and TCK while referring to the feature of JDK21, but without using the APIs of JDK21?

**Comment by Emily-Jiang** [comment]
> > Even Jakarta Concurrency which has support for creating work on virtual threads doesn't require Java SE 21 in the API and support for virtual threads is essentially optional.
> 
> Even if it is optional, as long as it exists in the spec, we will need an implementation and TCK to ensure that the API works correctly.

Just to clarify that the Virtual Thread support in Jakarta Concurrency is NOT optional. The API is just to set the thread to virtual or not. The Javadoc for this method documents clearly and mention it is kind of best effort if possible (See [here](https://github.com/jakartaee/concurrency/blob/main/api/src/main/java/jakarta/enterprise/concurrent/ManagedThreadFactoryDefinition.java#L190) for more details.). If an app is running in Java 17, obviously the Virtual Thread support has no effect. If an app is running on Java 21, it will be truly virtual if you want the thread to be virtual. Obviously, you can also set the thread to not virtual, platform thread. 

As for the TCKs, Nathan Rauh from Jakarta Concurrency said the tests would be written via reflection without directly depending on Java 21, thread.isThread(). In this way, either Java 17 or 21 can pass the tests. Obviously the behavior would be different for Java 17 and Java 21. 
Hope this helps!

**Comment by smillidge** [comment]
Given the issue raised is really about market adoption I think this should be raised through the Steering committee rather than decided at the project level. Given the strong steer towards JDK 21 as a baseline from the steering committee.

**Comment by JanWesterkamp-iJUG** [comment]
There is some adoption data from the Adoptium project, thats gives a little bit different result, espcially compared to the NewRelic report:

https://dash.adoptium.net/

There is also a trent view, that shows current data:

https://dash.adoptium.net/trends
